shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,88 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
-            {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, genericArguments.Length, fullName);
+        }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+        private static void ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int limit, bool fullName)
+        {
+            var offset = 0;
+            var nestingType = type.DeclaringType;
+
+            if (nestingType != null)
+            {
+                offset = nestingType.GetGenericArguments().Length;
             }
-            for (var i = 0; i < totalParts; i++)
+
+            if (fullName)
             {
-                var part = parts[i];
-                var num = part.IndexOf('`');
-                if (num != -1)
+                if (nestingType != null)
                 {
-                    var name = part.Substring(0, num);
-                    var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                    if (fullName || i == totalParts - 1)
-                    {
-                        sb.Append(name);
-                        AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                    }
-                    if (fullName && i != totalParts - 1)
-                    {
-                        sb.Append("+");
-                    }
-                    index += numberOfGenericTypeArgs;
+                    ProcessNestedGenericType(builder, nestingType, genericArgs, offset, fullName);
+                    builder.Append('+');
                 }
                 else
                 {
-                    if (fullName || i == totalParts - 1)
-                    {
-                        sb.Append(part);
-                    }
-                    if (fullName && i != totalParts - 1)
-                    {
-                        sb.Append("+");
-                    }
+                    builder.Append(type.Namespace);
+                    builder.Append('.');
+                }
+            }
+
+            var genericPartIndex = -1;
+            if (type.IsGenericType)
+            {
+                genericPartIndex = type.Name.IndexOf('`');
+            }
+
+            if (genericPartIndex > 0)
+            {
+                builder.Append(type.Name, 0, genericPartIndex);
+            }
+            else
+            {
+                builder.Append(type.Name);
+            }
+
+            if (offset == limit)
```

Comment:
If `` ` `` is found then type is generic and it will have some args to print. With correct math, it will never cross boundary.

Reply:
`Outer<int>.Inner` also generic type, and we don't need to print generic args for it.

Reply:
If no backtick found then we can return.